core/arch/arm/mm/tee_pager.c

```diff
@@ -201,84 +211,124 @@ static void pager_unlock(uint32_t exceptions)
+			idx++;
+		}
 
-	for (idx = core_mmu_va2idx(ti, CFG_TEE_RAM_START);
-	     idx < end_idx; idx++) {
-		core_mmu_get_entry(ti, idx, &p, &a);
-		if ((a & TEE_MATTR_VALID_BLOCK) && p == pa)
-			return (void *)core_mmu_idx2va(ti, idx);
+		n++;
+		if (n >= ARRAY_SIZE(pager_tables))
+			return NULL;
+		idx = 0;
 	}
 
 	return NULL;
 }
 
+static struct pager_table *find_pager_table_may_fail(vaddr_t va)
+{
+	size_t n;
+	const vaddr_t mask = CORE_MMU_PGDIR_MASK;
+
+	n = ((va & ~mask) - pager_tables[0].tbl_info.va_base) >>
+	    CORE_MMU_PGDIR_SHIFT;
+	if (n >= ARRAY_SIZE(pager_tables))
+		return NULL;
+
+	assert(va >= pager_tables[n].tbl_info.va_base &&
+	       va <= (pager_tables[n].tbl_info.va_base | mask));
+
+	return pager_tables + n;
+}
+
+static struct pager_table *find_pager_table(vaddr_t va)
+{
+	struct pager_table *pt = find_pager_table_may_fail(va);
+
+	if (!pt)
+		panic();
+	return pt;
+}
+
 bool tee_pager_get_table_info(vaddr_t va, struct core_mmu_table_info *ti)
 {
-	if (va >= (CFG_TEE_LOAD_ADDR & ~CORE_MMU_PGDIR_MASK) &&
-	    va <= (CFG_TEE_LOAD_ADDR | CORE_MMU_PGDIR_MASK)) {
-		*ti = tee_pager_tbl_info;
-		return true;
-	}
+	struct pager_table *pt = find_pager_table_may_fail(va);
 
-	return false;
+	if (!pt)
+		return false;
+
+	*ti = pt->tbl_info;
+	return true;
+}
+
+static struct core_mmu_table_info *find_table_info(vaddr_t va)
+{
+	return &find_pager_table(va)->tbl_info;
+}
+
+static struct pgt *find_core_pgt(vaddr_t va)
+{
+	return &find_pager_table(va)->pgt;
 }
 
 static void set_alias_area(tee_mm_entry_t *mm)
 {
-	struct core_mmu_table_info *ti = &pager_alias_tbl_info;
-	size_t tbl_va_size;
+	struct pager_table *pt;
 	unsigned idx;
-	unsigned last_idx;
 	vaddr_t smem = tee_mm_get_smem(mm);
 	size_t nbytes = tee_mm_get_bytes(mm);
+	vaddr_t v;
 
 	DMSG("0x%" PRIxVA " - 0x%" PRIxVA, smem, smem + nbytes);
 
-	if (pager_alias_area)
-		panic("null pager_alias_area");
-
-	if (!ti->num_entries && !core_mmu_find_table(smem, UINT_MAX, ti))
-		panic("Can't find translation table");
-
-	if ((1 << ti->shift) != SMALL_PAGE_SIZE)
-		panic("Unsupported page size in translation table");
-
-	tbl_va_size = (1 << ti->shift) * ti->num_entries;
-	if (!core_is_buffer_inside(smem, nbytes,
-				   ti->va_base, tbl_va_size)) {
-		EMSG("area 0x%" PRIxVA " len 0x%zx doesn't fit it translation table 0x%" PRIxVA " len 0x%zx",
-		     smem, nbytes, ti->va_base, tbl_va_size);
-		panic();
-	}
-
-	if (smem & SMALL_PAGE_MASK || nbytes & SMALL_PAGE_MASK)
-		panic("invalid area alignment");
-
 	pager_alias_area = mm;
 	pager_alias_next_free = smem;
 
 	/* Clear all mapping in the alias area */
-	idx = core_mmu_va2idx(ti, smem);
-	last_idx = core_mmu_va2idx(ti, smem + nbytes);
-	for (; idx < last_idx; idx++)
-		core_mmu_set_entry(ti, idx, 0, 0);
+	pt = find_pager_table(smem);
+	idx = core_mmu_va2idx(&pt->tbl_info, smem);
+	while (pt <= (pager_tables + ARRAY_SIZE(pager_tables) - 1)) {
+		while (idx < TBL_NUM_ENTRIES) {
+			v = core_mmu_idx2va(&pt->tbl_info, idx);
```

Comment:
This is a matter of taste. In this case I think the code is easier to read with v declared above together with the other local variables.